Source/WebCore/rendering/updating/RenderTreeUpdater.cpp

```diff
@@ -344,19 +344,22 @@ void RenderTreeUpdater::updateElementRenderer(Element& element, const Style::Ele
     }
 
     bool hasDisplayContents = elementUpdate.style->display() == DisplayType::Contents;
+    bool isSkippedContent = (element.isInTopLayer() || !m_document.isSkippedContentIgnored()) && renderTreePosition().parent().shouldSkipContent();
     if (hasDisplayContents)
         element.storeDisplayContentsStyle(makeUnique<RenderStyle>(WTFMove(elementUpdateStyle)));
+    else if (isSkippedContent)
+        element.storeSkippedContentStyle(makeUnique<RenderStyle>(WTFMove(elementUpdateStyle)));
     else
         element.clearDisplayContentsStyle();
```

Comment:
I think it might be better to move the check for `hasDisplayContents` into the definition of `isSkippedContent` and include the specification language and link to the spec:

// The specification https://drafts.csswg.org/css-contain/#size-containment says that
// content-visibility has no effect on an element:
// "if the element does not generate a principal box (as is the case with display: contents"

In addition, should `element.clearDisplayContentsStyle();` be called here even if the element skips its content?

Reply:
I think these are kind of independent things, i.e we need to keep track of both display:contents and is skipped content to make good decisions. The dependent thing is that we need to decide where the style goes to as we can only move it once. I decided to give display: contents priority for that, but it does mean I had to adapt Element::isSkippedContent slightly.

I did miss the element.clearDisplayContentsStyle() initially, well spotted!